Implement @font-face size-adjust descriptor (253062)

vitorroriz opened the pull request:
#### 97f82bc0e98c7ae32a6a0b24ded9ead1400583c8
<pre>
Implement @font-face size-adjust descriptor (253062)
<a href="https://bugs.webkit.org/show_bug.cgi?id=253062">https://bugs.webkit.org/show_bug.cgi?id=253062</a>
rdar://106349717

Reviewed by Myles C. Maxfield.

Implements @font-face size-adjust descriptor according to
<a href="https://www.w3.org/TR/css-fonts-5/#size-adjust-desc">https://www.w3.org/TR/css-fonts-5/#size-adjust-desc</a>

* LayoutTests/TestExpectations:
* Source/WebCore/css/FontFace.cpp:
(WebCore::FontFace::create):
(WebCore::FontFace::setSizeAdjust):
(WebCore::FontFace::sizeAdjust const):
* Source/WebCore/css/FontFace.h:
* Source/WebCore/css/FontFace.idl:
- Integrating size-adjust descriptor to FontFace() api.

* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
(WebCore::CSSPropertyParserWorkerSafe::parseFontFaceSizeAdjust):
* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.h:
- size-adjust parser wrapper for being used by FontFace().

* Source/WebCore/platform/graphics/FontPlatformData.cpp:
(WebCore::FontPlatformData::usedFontSize):
    It is not worth modifying the used size with @font-face size-adjust if
    we are to re-adjust it later with font-size-adjust.
This is because font-size-adjust will overrule this change, since
size-adjust also modifies the font&apos;s metric values and thus, keeps the
aspect-value unchanged (see discussion at w3c/csswg-drafts#6128).

* LayoutTests/TestExpectations:
* LayoutTests/platform/ios/size-adjust-interacts-with-text-size-adjust-auto-expected-mismatch.html: Added.
* LayoutTests/platform/ios/size-adjust-interacts-with-text-size-adjust-auto-notref.html: Added.
* LayoutTests/platform/ios/size-adjust-interacts-with-text-size-adjust-auto.html: Added.
* LayoutTests/platform/ios/size-adjust-interacts-with-text-size-adjust-percentage-expected.html: Added.
* LayoutTests/platform/ios/size-adjust-interacts-with-text-size-adjust-percentage-ref.html: Added.
* LayoutTests/platform/ios/size-adjust-interacts-with-text-size-adjust-percentage.html: Added.
- Theses tests won&apos;t be exported to WPT because -webkit-text-size-adjust
is a webkit-only property.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml

Canonical link: <a href="https://commits.webkit.org/263604@main">https://commits.webkit.org/263604@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e0e00d939a9385997969b75f58fa2861a6b792b1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5180 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5314 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5494 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6705 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/5253 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5177 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5523 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5291 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/5558 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5272 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5357 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4644 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6736 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/2843 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4642 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/10389 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/4291 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4707 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4715 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/6345 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/4769 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5144 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/4234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/5276 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4622 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/4638 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1330 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1242 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8705 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/5425 "Built successfully") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/4987 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1445 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
```diff
@@ -207,6 +207,23 @@ RefPtr<CSSPrimitiveValue> CSSPropertyParserWorkerSafe::parseFontFaceDisplay(cons
+RefPtr<CSSValue> CSSPropertyParserWorkerSafe::parseFontFaceSizeAdjust(const String& string, ScriptExecutionContext& context)
+{
+    if (!context.settingsValues().cssFontFaceSizeAdjustEnabled)
+        return nullptr;
```
I don't think you need this if you change the IDL as mentioned.

vitorroriz replied:
Great to know, thanks!

nt1m commented on Source/WebCore/css/FontFace.idl:
```diff
@@ -54,6 +55,7 @@ dictionary FontFaceDescriptors {
     [SetterCallWith=CurrentScriptExecutionContext] attribute DOMString unicodeRange;
     [SetterCallWith=CurrentScriptExecutionContext] attribute DOMString featureSettings;
     [SetterCallWith=CurrentScriptExecutionContext] attribute DOMString display;
+    [SetterCallWith=CurrentScriptExecutionContext] attribute DOMString sizeAdjust;
```
```suggestion
    [EnabledBySetting=CSSFontFaceSizeAdjust, SetterCallWith=CurrentScriptExecutionContext] attribute DOMString sizeAdjust;
```

litherum commented on Source/WebCore/platform/graphics/cocoa/FontFamilySpecificationCoreText.cpp:
```diff
@@ -48,6 +48,7 @@ FontFamilySpecificationCoreText::~FontFamilySpecificationCoreText() = default;
 
 FontRanges FontFamilySpecificationCoreText::fontRanges(const FontDescription& fontDescription) const
 {
+    // FIXME: do we also need to multiply size by @font-face sizeAdjust here?
```
Please add a test to answer this. :)

litherum commented on Source/WebCore/platform/graphics/freetype/FontCacheFreeType.cpp:
```diff
@@ -463,7 +463,7 @@ std::unique_ptr<FontPlatformData> FontCache::createFontPlatformData(const FontDe
 #endif
 
-    auto size = fontDescription.computedPixelSize();
+    float size = FontPlatformData::usedFontSize(fontDescription.computedPixelSize(), fontDescription.fontSizeAdjust(), fontCreationContext.sizeAdjust());
```
Why no `auto`?

litherum commented on Source/WebCore/platform/graphics/FontPlatformData.cpp:
```diff
@@ -106,6 +106,12 @@ const FontPlatformData::CreationData* FontPlatformData::creationData() const
+float FontPlatformData::usedFontSize(float computedFontSize, const FontSizeAdjust& fontSizeAdjust, float fontFaceSizeAdjust)
+{
+    // It is not worth modifying the used size with @font-face size-adjust if we are to re-adjust it later with font-size-adjust. This is because font-size-adjust will overrule this change, since size-adjust also modifies the font's metric values and thus, keeps the aspect-value unchanged. This allows us to use the used value as computed value later on updateSizeWithFontSizeAdjust(), since they w ...
+    return fontSizeAdjust.value ? computedFontSize : fontFaceSizeAdjust * computedFontSize;
```
What happens when the `font-size-adjust` property and the `font-size-adjust` descriptor disagree? Who wins?

How does this interact with -webkit-text-size-adjust (on both iPhone and iPad, which use different algorithms?) I think we need some WebKit-specific tests for these interactions.

vitorroriz replied:
`font-size-adjust` should win because `size-adjust` doesn't affect the aspect ratio, and font-size-adjust will act on the computed font size, which is not altered.

shivamidow replied:
> What happens when the font-size-adjust property and the font-size-adjust descriptor disagree? Who wins?

That is clarified by the level 5 spec. https://www.w3.org/TR/css-fonts-5/#font-resources
> Note: ... Consequently, applying font-size-adjust and size-adjust together means that size-adjust appears to have no effect.

It might be better to have the pointer in the comment here.

shivamidow commented on Source/WebCore/platform/graphics/FontPlatformData.h:
```diff
@@ -166,6 +166,7 @@ class FontPlatformData {
 
     static FontPlatformData cloneWithSize(const FontPlatformData&, float);
     void updateSizeWithFontSizeAdjust(const FontSizeAdjust&);
+    static float usedFontSize(float, const FontSizeAdjust&, float);
```
/* Unofficial Review */
I suggest moving this static function to FontDescription as a method with a new name. e.g., `float FontDescription::adjustedFontSize(float sizeAdjust);` Then, we don't need to pass two arguments of FontSizeAdjust to this function.

vitorroriz replied:
That's a good idea, will do, thanks

vitorroriz replied:
However, the name `adjusted` might be confusing because this is just the adjusted size regarding @font-face size-adjust, but not regarding font-size-adjust. I'm thinking to call it `adjustedSizeForFontFace` although sounds a bit silly.

shivamidow replied:
I am not good at naming. :P Maybe, `sizeAdjustedByFontFace`?

shivamidow commented on Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
```diff
@@ -673,7 +673,7 @@ static void autoActivateFont(const String& name, CGFloat size)
 std::unique_ptr<FontPlatformData> FontCache::createFontPlatformData(const FontDescription& fontDescription, const AtomString& family, const FontCreationContext& fontCreationContext)
 {
-    float size = fontDescription.computedPixelSize();
+    float size = FontPlatformData::usedFontSize(fontDescription.computedPixelSize(), fontDescription.fontSizeAdjust(), fontCreationContext.sizeAdjust());
```
/* Unofficial Review */
If FontDescription has the method, simply `auto size = fontDescription.adjustedFontSize(fontCreationContext.sizeAdjust());`

nt1m commented on LayoutTests/TestExpectations:
```diff
@@ -3722,11 +3722,10 @@ webkit.org/b/219735 imported/w3c/web-platform-tests/css/css-fonts/font-face-unic
 # @font-face: size-adjust
-webkit.org/b/206881 imported/w3c/web-platform-tests/css/css-fonts/size-adjust-01.html [ ImageOnlyFailure ]
+#  Flaky behavior
 webkit.org/b/206881 imported/w3c/web-platform-tests/css/css-fonts/size-adjust-text-decoration.tentative.html [ ImageOnlyFailure ]
```
Is this flaky or just failing on GTK? If it's the later, please put the expectation in LayoutTests/platform/gtk/TestExpectations or LayoutTests/platform/glib/TestExpectations (if this affects both WPE & GTK)

vitorroriz committed:
Implement @font-face size-adjust descriptor (253062)
https://bugs.webkit.org/show_bug.cgi?id=253062
rdar://106349717

Reviewed by Myles C. Maxfield.

Implements @font-face size-adjust descriptor according to
https://www.w3.org/TR/css-fonts-5/#size-adjust-desc

* LayoutTests/TestExpectations:
* Source/WebCore/css/FontFace.cpp:
(WebCore::FontFace::create):
(WebCore::FontFace::setSizeAdjust):
(WebCore::FontFace::sizeAdjust const):
* Source/WebCore/css/FontFace.h:
* Source/WebCore/css/FontFace.idl:
- Integrating size-adjust descriptor to FontFace() api.

* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
(WebCore::CSSPropertyParserWorkerSafe::parseFontFaceSizeAdjust):
* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.h:
- size-adjust parser wrapper for being used by FontFace().

* Source/WebCore/platform/graphics/FontPlatformData.cpp:
(WebCore::FontPlatformData::usedFontSize):
    It is not worth modifying the used size with @font-face size-adjust if
    we are to re-adjust it later with font-size-adjust.
This is because font-size-adjust will overrule this change, since
size-adjust also modifies the font's metric values and thus, keeps the
aspect-value unchanged (see discussion at w3c/csswg-drafts#6128).

* LayoutTests/TestExpectations:
* LayoutTests/platform/ios/size-adjust-interacts-with-text-size-adjust-auto-expected-mismatch.html: Added.
* LayoutTests/platform/ios/size-adjust-interacts-with-text-size-adjust-auto-notref.html: Added.
* LayoutTests/platform/ios/size-adjust-interacts-with-text-size-adjust-auto.html: Added.
* LayoutTests/platform/ios/size-adjust-interacts-with-text-size-adjust-percentage-expected.html: Added.
* LayoutTests/platform/ios/size-adjust-interacts-with-text-size-adjust-percentage-ref.html: Added.
* LayoutTests/platform/ios/size-adjust-interacts-with-text-size-adjust-percentage.html: Added.
- Theses tests won't be exported to WPT because -webkit-text-size-adjust
is a webkit-only property.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml

Canonical link: https://commits.webkit.org/263604@main
